### ISSUE TITLE
Any plans on a release?

### ISSUE THREAD

**Issue by polyzen** [himdel/hsetroot#7]

**Comment by himdel** [comment]
Well, I'm not opposed to it, but.. what for? :)

As in, are you looking for a specific distro release?
Or any specific features?
Or would you just like the current state tagged with a number?

**Comment by polyzen** [comment]
> Or would you just like the current state tagged with a number?

I probably first found your repo when looking into hsetroot for use with Compton, and had started watching it some time ago. More recently I took over packaging of hsetroot for Arch Linux, and began looking for how to track possible releases. https://directory.fsf.org/wiki/Hsetroot pointed me to the Debian releases, but eventually I found my way back here.

**Comment by himdel** [comment]
Aah, could it be that this fork is the last maintained version of hsetroot? I can certainly put some effort into making it into a proper release :).

Do you have a checklist of things you'd need?

**Comment by polyzen** [comment]
I thought so, but haven't yet tested. Great!

**Comment by polyzen** [comment]
When attempting to build [himdel-hsetroot-git](https://github.com/himdel/hsetroot/blob/master/pkg/arch/PKGBUILD.git): https://ptpb.pw/uplP

Also a [pkgver function](https://wiki.archlinux.org/index.php/VCS_package_guidelines#The_pkgver.28.29_function) should be used. On a side note, you may want to tag the initial commit as `1.0.2`.

**Comment by himdel** [comment]
@polyzen Feel free to open a pull request against the repo, I don't know anything about arch, and the PKGBUILD file was contributed by somebody else.

But.. from what I can tell, the pkgbuild file should be good, I can compile it locally just fine using just `make`:

```
himdel@thror:~/hsetroot$ make
cc -g -O2 -Wall     -c -o hsetroot.o hsetroot.c
cc -g -O2 -Wall     -c -o outputs_xrandr.o outputs_xrandr.c
cc  -lX11 -lImlib2  hsetroot.o outputs_xrandr.o   -o hsetroot

himdel@thror:~/hsetroot$ $ ldd ./hsetroot
	linux-vdso.so.1 (0x00007ffdb0b11000)
	libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6 (0x00007fbc64939000)
	libImlib2.so.1 => /usr/lib/x86_64-linux-gnu/libImlib2.so.1 (0x00007fbc646d5000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fbc64514000)
	libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1 (0x00007fbc644ea000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fbc644e5000)
	libfreetype.so.6 => /usr/lib/x86_64-linux-gnu/libfreetype.so.6 (0x00007fbc64429000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fbc642a4000)
	libXext.so.6 => /usr/lib/x86_64-linux-gnu/libXext.so.6 (0x00007fbc64092000)
	libX11-xcb.so.1 => /usr/lib/x86_64-linux-gnu/libX11-xcb.so.1 (0x00007fbc6408d000)
	libxcb-shm.so.0 => /usr/lib/x86_64-linux-gnu/libxcb-shm.so.0 (0x00007fbc64088000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fbc64a9d000)
	libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6 (0x00007fbc63e84000)
	libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6 (0x00007fbc63c7e000)
	libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16 (0x00007fbc63c43000)
	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fbc63a25000)
	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007fbc6380e000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fbc63804000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fbc637e3000)
```

That's on debian, using these versions (amd64 architecture):

libx11-6 2:1.6.7-1
libimlib2 1.5.1-1

**Comment by polyzen** [comment]
On January 10, 2019 5:00:15 AM EST, Martin Hradil <notifications@github.com> wrote:
>@polyzen Feel free to open a pull request against the repo, I don't
>know anything about arch, and the PKGBUILD file was contributed by
>somebody else.

Will do 

>But.. from what I can tell, the pkgbuild file should be good, I can
>compile it locally just fine using just `make`:
>
>```
>himdel@thror:~/hsetroot$ make
>cc -g -O2 -Wall     -c -o hsetroot.o hsetroot.c
>cc -g -O2 -Wall     -c -o outputs_xrandr.o outputs_xrandr.c
>cc  -lX11 -lImlib2  hsetroot.o outputs_xrandr.o   -o hsetroot
>
>himdel@thror:~/hsetroot$ $ ldd ./hsetroot
>	linux-vdso.so.1 (0x00007ffdb0b11000)
>	libX11.so.6 => /usr/lib/x86_64-linux-gnu/libX11.so.6
>(0x00007fbc64939000)
>	libImlib2.so.1 => /usr/lib/x86_64-linux-gnu/libImlib2.so.1
>(0x00007fbc646d5000)
>	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fbc64514000)
>	libxcb.so.1 => /usr/lib/x86_64-linux-gnu/libxcb.so.1
>(0x00007fbc644ea000)
>	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fbc644e5000)
>	libfreetype.so.6 => /usr/lib/x86_64-linux-gnu/libfreetype.so.6
>(0x00007fbc64429000)
>	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fbc642a4000)
>	libXext.so.6 => /usr/lib/x86_64-linux-gnu/libXext.so.6
>(0x00007fbc64092000)
>	libX11-xcb.so.1 => /usr/lib/x86_64-linux-gnu/libX11-xcb.so.1
>(0x00007fbc6408d000)
>	libxcb-shm.so.0 => /usr/lib/x86_64-linux-gnu/libxcb-shm.so.0
>(0x00007fbc64088000)
>	/lib64/ld-linux-x86-64.so.2 (0x00007fbc64a9d000)
>	libXau.so.6 => /usr/lib/x86_64-linux-gnu/libXau.so.6
>(0x00007fbc63e84000)
>	libXdmcp.so.6 => /usr/lib/x86_64-linux-gnu/libXdmcp.so.6
>(0x00007fbc63c7e000)
>	libpng16.so.16 => /usr/lib/x86_64-linux-gnu/libpng16.so.16
>(0x00007fbc63c43000)
>	libz.so.1 => /lib/x86_64-linux-gnu/libz.so.1 (0x00007fbc63a25000)
>	libbsd.so.0 => /lib/x86_64-linux-gnu/libbsd.so.0 (0x00007fbc6380e000)
>	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fbc63804000)
>	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0
>(0x00007fbc637e3000)
>```
>
>That's on debian, using these versions (amd64 architecture):
>
>libx11-6 2:1.6.7-1
>libimlib2 1.5.1-1

The issue occurs with the added hardening flags enforced by Arch.


--
Best,
polyzen

**Comment by himdel** [comment]
> On a side note, you may want to tag the initial commit as 1.0.2.

Pushed :)

**Comment by himdel** [comment]
> The issue occurs with the added hardening flags enforced by Arch.

Any idea what those are?

I can try fixing that if I can reproduce it, but, all I found is `-fno-plt -fstack-check` and `-z *something*`, but neither seems to break the build for me (have not tried all the variants of something).

**Comment by himdel** [comment]
Oh, that's a problem :( Looks like arch may be using an entirely different linker?

(I'm getting:

```
cc: error: unrecognized command line option ‘-Wl’; did you mean ‘-W’?
cc: error: unrecognized command line option ‘--sort-common’; did you mean ‘--no-common’?
cc: error: unrecognized command line option ‘--as-needed’
```
)

**Comment by polyzen** [comment]
Perhaps if you specify gcc? Based on the Debian hardening docs, those flags should be possible (though I did not see `--sort-common`.

**Comment by himdel** [comment]
Aah, sorry, my fault.. I parsed it as a comma-separated list of args.

It's actually one long arg, for some reason:

```
cc  -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -lX11 -lImlib2  hsetroot.o outputs_xrandr.o   -o hsetroot
/usr/bin/ld: hsetroot.o: in function `setRootAtoms':
/home/himdel/hsetroot/hsetroot.c:75: undefined reference to `XInternAtom'
/usr/bin/ld: /home/himdel/hsetroot/hsetroot.c:76: undefined reference to `XInternAtom'
...
/usr/bin/ld: /home/himdel/hsetroot/hsetroot.c:564: undefined reference to `XFreePixmap'
collect2: error: ld returned 1 exit status
```

(about a hundred lines of errors)

**Comment by himdel** [comment]
Looks like one immediate solution is to disable `--as-needed`.

**Comment by himdel** [comment]
Adding

```
LDFLAGS+=-Wl,--no-as-needed
```

should help :).

**Comment by himdel** [comment]
If I'm reading the `as-needed` docs correctly, it looks like just adding the missing libraries would also solve the problem.

But, apparently not, adding `-lc -lxcb -ldl -lfreetype -lm -lXext -lXau -lXdmcp -lpng16 -lz -lrt -lpthread`  doesn't do a thing.

--

So, I guess `--no-as-needed` wins :)

**Comment by polyzen** [comment]
Thank you! Will see if this appeases `checksec` and investigate further.

**Comment by polyzen** [comment]
Looks good to me (all green):
```
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      Symbols         FORTIFY Fortified       Fortifiable  FILE
Full RELRO      Canary found      NX enabled    PIE enabled     No RPATH   No RUNPATH   No Symbols       Yes    2               3       hsetroot 
```

**Comment by himdel** [comment]
OK, released current state as 1.0.3, so I guess we can close this :)
